### ISSUE TITLE
Doesn't have write permission on internal storage

### ISSUE THREAD

**Issue by podhorsky-ksj** [wolpi/prim-ftpd#171]
Mode "Plain old filesystem" access is only read only and other non-root accesses doesn't work on internal storage of android 8.1

I check on setting app have all permission it requires.

**Comment by podhorsky-ksj** [comment]
I have SD card and on it SAF is ok and I can connect to it via ftp. But when I choose internal storage, ftpd print no error (as with SD card), but I cannot connect. 
Client Free file sync gives me this error:
"CURLE_COULDNT_CONNECT: Failed to connect to 192.168.43.1 port 12345: Connection refused [curl_easy_perform]"

I tried android app FTP Server (Free) and it works ok on internal storage, so I think there is an issue in prim-ftpd app.

**Comment by wolpi** [comment]
> Failed to connect to 192.168.43.1 port 12345: Connection refused 

Usually that means that server was not started. Please double check if server ist really running.

Logs would still bei great :wink:

**Comment by podhorsky-ksj** [comment]
Attaching log file. According to log ftpd is running.
[prim-ftpd-log-2019-01-22-10-01-57.zip](https://github.com/wolpi/prim-ftpd/files/2782055/prim-ftpd-log-2019-01-22-10-01-57.zip)

**Comment by wolpi** [comment]
Yes, server is running and IP and Port from above are valid. 

It makes me wonder that there are quite some "network connectivity changed" events in your log.

The client error `Connection refused` means that client cannot reach server. As we know from logs that server is OK it seems like a network issue.

**Comment by podhorsky-ksj** [comment]
No, network is ok, when i try it with another app or try to put saf on sd card, everything is ok.
I think only remains issue in this app. But I don't know which.

**Comment by podhorsky-ksj** [comment]
Yes, there are some odd wifi reconnection. But these permissions returns different result code

In mode "Plain old filesystem"
In client: 
`
CURLE_QUOTE_ERROR: QUOT command failed with 450
450 No permission to delete /DocumentsShared/2019-01-20_12-17 - foldersync.db. [curl_easy_perform]
`
In server
`
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a MESSAGE_RECEIVED event for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.filter.codec.ProtocolCodecFilter";"Processing a MESSAGE_RECEIVED for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a MESSAGE_RECEIVED event for session 3";"";
"2019-01-26 10:06:10.3";"INFO";"org.apache.ftpserver.listener.nio.FtpLoggingFilter";"RECEIVED: DELE /DocumentsShared/2019-01-20_12-17 - foldersync.db";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFileSystemView";"getFile(/DocumentsShared/2019-01-20_12-17 - foldersync.db)";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFile";"[0] getAbsolutePath() -> '/storage/emulated/0'";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFileSystemView";"  finding abs path for '/DocumentsShared/2019-01-20_12-17 - foldersync.db' with wd '/storage/emul...";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFile";"[0] getAbsolutePath() -> '/storage/emulated/0'";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFileSystemView";"  getFile(abs: /DocumentsShared/2019-01-20_12-17 - foldersync.db)";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFile";"[2019-01-20_12-17 - foldersync.db] getAbsolutePath() -> '/DocumentsShared/2019-01-20_12-17 - fold...";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFile";"[2019-01-20_12-17 - foldersync.db] isDirectory() -> false";"";
"2019-01-26 10:06:10.3";"TRACE";"org.primftpd.filesystem.FsFtpFile";"[2019-01-20_12-17 - foldersync.db] isRemovable()";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a WRITE event for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a WRITE event for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event WRITE has been fired for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a WRITE event for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event WRITE has been fired for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event WRITE has been fired for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a MESSAGE_SENT event for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event MESSAGE_RECEIVED has been fired for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.filter.executor.OrderedThreadPoolExecutor";"Adding event MESSAGE_SENT to session 3
Queue : [MESSAGE_SENT, ]
";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event MESSAGE_SENT has been fired for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event MESSAGE_RECEIVED has been fired for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a MESSAGE_SENT event for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Firing a MESSAGE_SENT event for session 3";"";
"2019-01-26 10:06:10.3";"INFO";"org.apache.ftpserver.listener.nio.FtpLoggingFilter";"SENT: 450 No permission to delete /DocumentsShared/2019-01-20_12-17 - foldersync.db.
";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event MESSAGE_SENT has been fired for session 3";"";
"2019-01-26 10:06:10.3";"DEBUG";"org.apache.mina.core.filterchain.IoFilterEvent";"Event MESSAGE_SENT has been fired for session 3";"";
`

**Comment by wolpi** [comment]
Well, it is likely that file `/DocumentsShared/2019-01-20_12-17 - foldersync.db` is really not writable/deletable.
You may open a shell and check permission with `ls -l`.

**Comment by podhorsky-ksj** [comment]
-rw-rw----    1 root     everybod    110592 Jan 21 07:43 2019-01-21_07-43 - foldersync.db

It has permissions like the other files. App probably doesn't have such permissions, but i gave app all permissions, app asked for. Other ftp server app transfer files without issue.

**Comment by wolpi** [comment]
>-rw-rw---- 1 root everybod 110592 Jan 21 07:43 2019-01-21_07-43 - foldersync.db

Those linux permission indicate the file should be writable by group `everyone`. Not shure whether that includes this app. Even if it woulbe be included, the system obviously blocks write access anyway. There are many reports about weird system behavior regarding filesystem access, here are some: #31, #51, #65, #67, #106, #145.

Android permissions and linux permissions are two different worlds. An android app can just request permisson to access filesystem. But that is restricted to a certain path. Nowadays it is usually `/storage/emulated/0`. As you wrote "internal storage" I assumed you were refering to that. Thus I gave missleading responses.

Many people ask about access to sdcard. Android has more issues with that. Apps should not access it via filesystem API anymore. Google invented an own API called SAF for that.

Long story short: you cannot access that path with this app. It is blocked by the operating system.

**Comment by podhorsky-ksj** [comment]
By internal storage i really mean /storage/emulated/0 this storage is normally accessible by all app, but this app doesn't have write access to it. The file i posted is in subdirectory od this path. It is accesible a writable by all apps in my android except this app. It is not blocked by android. 
I know permissions in linux a i think when android give write permission to some app, in linux meaning it is put app in a group in this case everybody. Your app probably didn't ask for such permission.

**Comment by wolpi** [comment]
Finally this starts making sense :laughing: 

According to your log you use storage type `plain old filesystem`. This means your path is wrong. It must be:
`/storage/emulated/0/DocumentsShared/2019-01-20_12-17 - foldersync.db`

Things are different with `SAF`. In that case you must choose a directory which is "exported" via ftp/sftp as `/`. 

You can see that on your own when browsing remote filesystem with a client.

**Comment by podhorsky-ksj** [comment]
I just tried that. when I set "plain old filesystem" and /, it will not find any of my files on mobile.
when i set "plain old filesystem" and "/storage/emulated/0" it will find my files, but they are read-only your app cannot modify them
Via SAF on /storage/emulated/0, it cannot cannot connect.
Via SAF on sd card it works correctly.

I don't want to test this app anymore. I found app which works for internal and sd card via ftp and I can modify and download my files. I cannot use ftpd on android 8.1 without some fix.

**Comment by wolpi** [comment]
We would run in circles. Now I would need logs showing your errors ...

I'm closing this now.
